File: 임현규/programmers/Lv1/lv1_키패드누르기_하.py
def solution(numbers, hand):
    def getDistance(num1, num2):
        y1, x1 = divmod(num1-1, 3)
        y2, x2 = divmod(num2-1, 3)
        return abs(y1 - y2) + abs(x1 - x2)
    
    result = []
    left = 10
    right = 12
    
    for num in numbers:
        if num == 0:
            num = 11
        if num in [1, 4, 7]:
            left = num
            result.append('L')
        elif num in [3, 6, 9]:
            right = num
            result.append('R')
        else: # 2, 5, 8, 0
            dLeft = getDistance(num, left)
            dRight = getDistance(num, right)
            if dLeft < dRight:
                left = num
                result.append('L')
            elif dLeft > dRight:
                right = num
                result.append('R')
            else: # 같은 경우
                if hand == 'left':
                    left = num
                    result.append('L')
                else:
                    right = num
                    result.append('R')
    return ''.join(result)

File: 임현규/programmers/Lv1/lv1_키패드 누르기_하.js
function solution(numbers, hand) {
    var handPosition = {
        left: "*",
        right: "#",
    };

    const answer = numbers.map(element => {
        if ([1, 4, 7].includes(element)) {
            handPosition["left"] = element;
            return "L";
        }
        if ([3, 6, 9].includes(element)) {
            handPosition.right = element;
            return "R";
        }

        if ([2, 5, 8, 0].includes(element)) {
            const val1 = calculateDistance(element, handPosition["left"]);
            const val2 = calculateDistance(element, handPosition["right"]);
            if (val1 > val2) {
                handPosition["right"] = element;
                return "R";
            }
            if (val1 < val2) {
                handPosition["left"] = element;
                return "L";
            }
            if (val1 == val2) {
                handPosition[hand] = element;
                return hand[0].toUpperCase();
            }
        }
    });
    return answer.join("");
}

function convert(pad) {
    const convertMap = {
        "*": 10,
        0: 11,
        "#": 12,
    };

    if (convertMap.hasOwnProperty(pad)) {
        return convertMap[pad];
    }
    return pad;
}

function getPlane(number) {
    const row = Math.floor((number - 1) / 3);
    const col = (number - 1) % 3;
    return [row, col];
}

function calculateDistance(a, b) {
    const [r1, c1] = getPlane(convert(a));
    const [r2, c2] = getPlane(convert(b));
    return Math.abs(r1 - r2) + Math.abs(c1 - c2);
}

console.log(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"));
